### PR TITLE
13 implement routes for getting the reviews

### LINKED ISSUE #13: Implement routes for getting the reviews
## Description
For the reviews page, we need to get every review of a user with the following data - 

- Rating
- Text
- Words to mark in the text (proof of sentiment)
- Sentiment
- Topic
- Data Source

## DoD

- [x] We have `/reviews` route
- [x] Route supports pagination

### PR DESCRIPTION
**Pull request by Shaked-abdu**
closes #13

### CODE REVIEW

File: backend/src/controllers/review.controller.ts [hunk 1]
```diff
@@ -29,6 +41,34 @@ class ReviewController extends BaseController<IReview> {
...
+  }
+
+  async getPaginated(req: AuthRequest<{}, {}, {}, IGetReviewsParams>, res: Response<IGetAllReviewsResponse>) {
+    const { limit, page, before, search } = req.query;
+
+    try {
+      const reviews = (await this.model
+        .find({ businessId: req.user!.businessId, date: { $lt: before } })
+        .limit(Number(limit))
+        .skip(Number(page - 1) * Number(limit))) as IReview[];
```
Comment (Shaked-abdu) [inline backend/src/controllers/review.controller.ts]: @liadyacobi what the "search" param syntax in the /reviews path?

http://localhost:8000/api/reviews?page=2&limit=10&before=2024-06-30T19:35:38.917Z&search=<?>

In order to know how to use it in the model query

File: backend/src/config.ts [hunk 2]
```diff
@@ -5,9 +5,7 @@ env.config();
 
 const ConfigSchema = z.object({
   datalakeDBUrl: z.string(),
-  datalakeDBName: z.string(),
```
Comment (liadyacobi) [inline backend/src/config.ts:8]: why did you remove these?

Comment (liadyacobi) [inline backend/src/config.ts:8]: why did you remove these?

File: backend/src/controllers/review.controller.ts (hunk 1 above)
Comment (liadyacobi) [reply backend/src/controllers/review.controller.ts]: the search is just of a `string` type. it should be typed in the IDE and parsed to string correctly

File: backend/src/controllers/review.controller.ts [hunk 3]
```diff
@@ -29,6 +41,34 @@ class ReviewController extends BaseController<IReview> {
...
+    });
+  }
+
+  async getPaginated(req: AuthRequest<{}, {}, {}, IGetReviewsParams>, res: Response<IGetAllReviewsResponse>) {
+    const { limit, page, before, search } = req.query;
+
+    try {
+      const reviews = (await this.model
+        .find({ businessId: req.user!.businessId, date: { $lt: before } })
+        .limit(Number(limit))
```
Comment (liadyacobi) [inline backend/src/controllers/review.controller.ts]: Do you require to convert these using `Number`? it should be typed correctly

File: backend/src/controllers/review.controller.ts [hunk 4]
```diff
@@ -29,6 +41,34 @@ class ReviewController extends BaseController<IReview> {
...
+        .skip(Number(page - 1) * Number(limit))) as IReview[];
+
+      return res.status(httpStatus.OK).send({
+        currentPage: Number(page),
+        totalReviews: await this.model.countDocuments({ businessId: req.user!.businessId }),
```
Comment (liadyacobi) [inline backend/src/controllers/review.controller.ts:67]: maybe we should set an index on the `businessId` property in the collection

File: backend/src/controllers/review.controller.ts [hunk 5]
```diff
@@ -29,6 +41,34 @@ class ReviewController extends BaseController<IReview> {
...
+        reviews,
+      });
+    } catch (error) {
+      return res.status(httpStatus.INTERNAL_SERVER_ERROR).send();
```
Comment (liadyacobi) [inline backend/src/controllers/review.controller.ts:72]: log the error

File: backend/src/db.ts [hunk 6]
```diff
@@ -4,10 +4,10 @@ import config from './config';
 
 const logger = createLogger('db');
 
-const initDB = async (dbUrl: string, dbName: string, mongoose: Mongoose) => {
-  const url = `mongodb://${dbUrl}/${dbName}`;
+const initDB = async (dbUrl: string, mongoose: Mongoose) => {
```
Comment (liadyacobi) [inline backend/src/db.ts:7]: Does it work without the DB name?

File: backend/src/routes/review.route.ts [hunk 7]
```diff
@@ -98,6 +98,68 @@ const router = Router();
...
+ *           description: The unique identifier of the review
+ *         value:
+ *           type: string
+ *           description: The review text
+ *         phrases:
+ *           type: array
+ *           items:
+ *             type: string
+ *             description: The phrase extracted from the review
+ *           description: The phrases extracted from the review
```
Comment (liadyacobi) [inline backend/src/routes/review.route.ts]: Duplicate description

**Review by liadyacobi — COMMENTED** [review]
Depends on #62

File: backend/src/config.ts (hunk 2 above)
Comment (Shaked-abdu) [reply backend/src/config.ts:8]: Cause we use port 27018 in the datalake DB (and 27017 in the metadatadb)

then we had hardcoded port 27017 (the default port of mongodb)

then
```
  const url = `mongodb://${dbUrl}:27017/${dbName}`;
```
now
```
  const url = `mongodb://${dbUrl}`;
```

note that dbUrl is for example: localhost:27018/datalake

File: backend/src/controllers/review.controller.ts (hunk 3 above)
Comment (Shaked-abdu) [reply backend/src/controllers/review.controller.ts]: copilot autocomplete, fixed

File: backend/src/db.ts (hunk 6 above)
Comment (Shaked-abdu) [reply backend/src/db.ts:7]: then
```
DATALAKE_DB_URL = localhost
DATALAKE_DB_NAME = datalake
METADATA_DB_URL = localhost
METADATA_DB_NAME = metadatadb
```

now
```
DATALAKE_DB_URL = localhost:27018/datalake
METADATA_DB_URL = localhost:27017/metadatadb
```

File: backend/src/controllers/review.controller.ts (hunk 4 above)
Comment (Shaked-abdu) [reply backend/src/controllers/review.controller.ts:67]: what do you mean?

**Comment by Shaked-abdu** [comment]
Merge only after #62 merged

File: backend/src/controllers/review.controller.ts (hunk 4 above)
Comment (liadyacobi) [reply backend/src/controllers/review.controller.ts:67]: In the model file we can define DB indexes. This will make queries like this much faster. see [here](https://mongoosejs.com/docs/guide.html#indexes)
